doc/GUIDE.md

```diff
@@ -0,0 +1,910 @@
+    // ...
+}
+```
+
+For working with arrays of structs, we can use the `allocateArray` method which accepts an
+additional element count, indicating the length of the array:
+
+```java
+// Main.java
+
+try (Arena arena = Arena.ofConfined()) {
+    int arrLen = 5;
+    MemorySegment points = Point.allocateArray(arrLen, arena);
+
+    for (int i = 0; i < arrLen; i++) {
+        MemorySegment element = Point.asSlice(points, i);
+        Point.x(element, 10 + i);
+        Point.y(element, 5 + i);
+    }
+
+    // ...
+}
+```
+
+In the above example, the `asSlice` method is used to _slice_ out a section of
+the array, which corresponds to a single `Point` struct element. This method
+can be used to access individual elements of the `points` array when given
+an index.
+
+Finally, the `reinterpret` method can be used to _sanitize_ a pointer that is
+returned from native code. Let's say we have a C function that creates an instance
+of a `Point`, and returns a pointer to it, as well as a function that deletes a
+point, given a pointer:
+
+```c
+struct Point* new_point(void);
+void delete_point(struct Point* ptr);
+```
+
+The pointer that is returned by the corresponding method that jextract generates
+for this function does not have the correct bounds or lifetime associated with it.
+These things are not possible to figure out automatically (for instance, a pointer
+could point at a single `Point` struct, or an array of multiple `Point` structs).
+
+The `reinterpret` method can be used to associate the correct bounds and lifetime:
+
+```java
+// Main.java
+
+try (Arena arena = Arena.ofConfined()) {
+    MemorySegment point = Point.reinterpret(new_point(), arena, mylib_h::delete_point);
+
+    // ...
+} // 'delete_point` called here
+```
+
+The `point` segment returned by `reinterpret` has exactly the size of one `Point` (for
+arrays of struct, use the `reinterpret` overload that takes an element count as well). The
+lifetime we associate with the segment is the lifetime denoted by `arena`, and when the
+arena is closed, we want to call `delete_point`, which we can do by passing a method
+reference to `delete_point` as a cleanup action when calling `reinterpret`.
+
+### Function Pointers
+
+Jextract generates a separate class for each function pointer type found in the header
+files it parses. For instance, for a function pointer `typedef` like this:
+
+```c
+// mylib.h
+
+typedef int (*callback_t)(int x, int y);
+```
+
+Jextract generates the following class in a `callback_t.java` file:
+
+```java
+// callback_t.java
+
+public class callback_t {
+    public interface Function {
+        int apply(int x, int y);
+    }
+
+    public static FunctionDescriptor descriptor() { ... }
+    public static MemorySegment allocate(callback_t.Function fi, Arena arena) { ... }
+    public static int invoke(MemorySegment funcPtr,int x, int y) { ... }
+}
+```
+
+We again have a meta-data accessor for the function descriptor (`descriptor()`). There's
```

Comment:
Maybe here using numbers in the code and a numbered list would be (a) more consistent and (b) more readable